Stop Using cs_main for CNodeState/State()

TheBlueMatt opened the pull request:
Based on #9243 and the top commit off of #9375 ("Make CBlockIndex*es in net_processing const"), this no longer requires cs_main to call State() and removes a few cs_main's around Misbehaving() for an easy win.

TheBlueMatt committed:
Make CBlockIndex*es in net_processing const

TheBlueMatt committed:
Move mapNodeState into an encapsulation class

TheBlueMatt committed:
Add a few AssertLockHelds for cs_main in net_processing

TheBlueMatt committed:
Optimize State() calls in ::SENDCMPCT processing

TheBlueMatt committed:
Remove cs_main lock requirements for Misbehaving()

TheBlueMatt committed:
Split CNode::cs_vSend: message processing and message sending

cs_vSend is used for two purposes - to lock the datastructures used
to queue messages to place on the wire and to only call
SendMessages once at a time per-node. I believe SendMessages used
to access some of the vSendMsg stuff, but it doesn't anymore, so
these locks do not need to be on the same mutex, and also make
deadlocking much more likely.

dcousens commented on src/util.cpp:
```diff
@@ -238,9 +240,12 @@ bool LogAcceptCategory(const char* category)
-            const vector<string>& categories = mapMultiArgs["-debug"];
-            ptrCategory.reset(new set<string>(categories.begin(), categories.end()));
-            // thread_specific_ptr automatically deletes the set when the thread ends.
+            if (mapMultiArgs.count("-debug")) {
```
Why not `IsMultiArgSet`?

TheBlueMatt replied:
Prefer less diff? (note this is really from #9243, not strictly this PR).

dcousens commented on src/util.cpp:
```diff
@@ -412,6 +428,13 @@ bool SoftSetBoolArg(const std::string& strArg, bool fValue)
 
+void ForceSetArg(const std::string& strArg, const std::string& strValue)
+{
+    mapArgs[strArg] = strValue;
```
Why is this not `LOCK(cs_args)`?

TheBlueMatt replied:
Fixed in #9243

sipa commented:
Wouldn't it be easier to use a shared_ptr wrapping for cleanup-after-last-use, instead of implementing refcounting yourself? You can still use a manual wrapper to lock/unlock.

TheBlueMatt commented:
Hmm, I kinda prefer to not introduce shared_ptr abstractions unless they're optimizing a copy that we don't want to do. I don't think the manual refcounting is that bad.

On December 27, 2016 7:46:52 PM GMT+01:00, Pieter Wuille <notifications@github.com> wrote:
>Wouldn't it be easier to use a shared_ptr wrapping for
>cleanup-after-last-use, instead of implementing refcounting yourself?
>You can still use a manual wrapper to lock/unlock.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/9419#issuecomment-269366047

TheBlueMatt committed:
Let shared_ptrs cleanup CNodeStates

This removes reliance on net in net_processing for maintaining the
refcount == 0 invariant when calling FinalizeNode

This also removes a few asserts which used to be checked when there
were no more CNodeStates remaining, which we can no longer check for.

These should be pretty rarely checked anyway, so probably didn't
serve much use.

TheBlueMatt commented:
OK, @sipa found a better way to use shared_ptrs here so I went ahead and did that. It means removing some asserts, but I think its worth it.

sipa commented on src/net_processing.cpp:
```diff
@@ -220,28 +223,65 @@ struct CNodeState {
+                return CNodeStateAccessor();
+            pstate = it->second;
+        }
+        return CNodeStateAccessor(pstate);
```
You should add an `std::move` here if you want to avoid a copy.

sipa commented on src/net_processing.cpp:
```diff
@@ -220,28 +223,65 @@ struct CNodeState {
+    std::shared_ptr<CNodeState> pstate;
+
+    CNodeStateAccessor() : pstate() { }
+    CNodeStateAccessor(std::shared_ptr<CNodeState> pstateIn) : pstate(pstateIn) { ENTER_CRITICAL_SECTION(pstate->cs); }
```
Add a `std::move` around `pstateIn` to avoid a copy.

TheBlueMatt commented:
Went down the rabbit hole here trying to track down a deadlock and then realized that it was already prevented. Note that (very much on purpose) cs_main is still used prior to all CNodeStateAccessors except for a few where we only call Misbehaving() or otherwise sure we are only locking one CNodeState at a time (to avoid deadlocks where we lock them in different orders).

TheBlueMatt committed:
Give CNodeState its own lock (and accessor class for RAII access)

TheBlueMatt commented:
Addressed @sipa's comments.

theuni commented on src/net_processing.cpp:
```diff
@@ -220,18 +223,68 @@ struct CNodeState {
+
+    CNodeStateAccessor(CNodeStateAccessor&& o) { pstate = o.pstate; o.pstate = NULL; }
+
+    bool operator()() const { return (bool)pstate; }
```
explicit operator bool instead? Using the call operator for this seems strange.

TheBlueMatt replied:
I dunno I dont really like the bool operator. If you really prefer it I'm happy to change it.

theuni replied:
explicit bool matches the semantics of smart pointers, so I think it's a good fit here. But more importantly, the call operator makes it look as though something is happening to the accessor, so I'd really rather not use that.

TheBlueMatt replied:
OK, switched to regular bool.

theuni commented on src/net_processing.cpp:
```diff
@@ -220,28 +223,65 @@ struct CNodeState {
+};
+
 class NodeStateStorage {
     /** Map maintaining per-node state. Requires cs_main. */
```
Stale comment?

TheBlueMatt replied:
Fixed.

theuni commented on src/net_processing.cpp:
```diff
@@ -220,28 +223,65 @@ struct CNodeState {
+    std::shared_ptr<CNodeState> pstate;
+
+    CNodeStateAccessor() : pstate() { }
+    CNodeStateAccessor(std::shared_ptr<CNodeState>&& pstateIn) : pstate(std::move(pstateIn)) { ENTER_CRITICAL_SECTION(pstate->cs); }
```
Why not just have a lock as a member?

TheBlueMatt replied:
I didnt realize we had an RAII lock in sync.h that would still do DEBUG_LOCKORDER checks. Will fix with CMutexLock.

TheBlueMatt replied:
Hmm, looking again using CMutexLock means adding __FILE__, __LINE__, etc macros in net_processing for parameters to CMutexLock, instead of letting ENTER_CRITICAL_SECTION figure it out. If you prefer I could try to add some crazy defines like CONSTRUCT_CMUTEXLOCK(mutex, lockName) which fills in a constructor for use in the CNodeStateAccessor constructor?

theuni replied:
Hmm, I think you actually want these coming from net_processing anyway, no? I think if you make the ctor inline and pass \_\_LINE__ to CMutexLock, you'd end up seeing the source of the lock. Otherwise, you're always seeing it come from the same place I should think.

Taking that a step further, I think it would work to change CMutexLock's ctors to take default \_\_FILE__ and \_\_LINE__ args?

theuni commented:
Wanted to make sure my nits so far were reasonable, so I went ahead and patched 'em in. Feel free to take from https://github.com/theuni/bitcoin/commit/058193e4d51bdd0cb8e7c225130414b817207ac2

TheBlueMatt commented:
Fixed a few of @theuni's comments, rebased diff-tree is at https://0bin.net/paste/nkpqzdxyoAXkoX7B#STuXxIS32GXf3gMP1JbiIPrufL9gq1hcSH+g6yXSKdq

theuni commented:
@TheBlueMatt Maybe I'm missing something big-picture here, but would it not be possible to hold a new global mutex rather than cs_main to serialize CNodeStateAccessor access? Or is there a reason that they must share the same lock?

sipa commented:
@theuni I think that's the plan longer term.

theuni commented:
I suppose I'm trying to figure out, for the sake of reviewing this PR, what remains to be done before that can happen. I'm certainly not insinuating that we should hold this up until that point, I'd just like to get an idea of what other issues still need to be solved.

So if, for ex, we naively made CNodeState::cs static and dropped the cs_main guards, what kind of carnage would be expected?

TheBlueMatt commented:
Aside from the places that do, actually, require cs_main, a global order between various CNodeState::cs locks needs to be defined (or, equivalently, we could say that you may not hold multiple CNodeState::cs locks at once).

TheBlueMatt commented:
Without #9488 (which I do not think should make 0.14 at this juncture), this should not go in for 0.14.

TheBlueMatt commented:
Closing for now. Will make access exclusive and work towards a more whole solution for 0.15.
